src/os/README.md

```diff
@@ -104,3 +108,125 @@ inspiration for several decisions in this library.
 - that said, if an operation is extremely efficient and almost always wanted,
   consider doing it natively, automatically. For example, ints in HSL
   `sockaddr` are always in host byte order, not network byte order.
+
+## Appendix:  C / Hack type mapping cheat sheet
+
+| C type | Hack type |
+|---|---|
+| `int` or `short` as a bit set | `keyset<Flag>` where `Flag` is an `enum` |
+| long-lived `int socket` or `int filedes` | `HH\Lib\OS\FileDescriptor` or other wrapper classes |
+| short-lived `void *` or `struct *` | `shape` or `vec`, see [Appendix: opaque C pointer encoding](#appendix-opaque-c-pointer-encoding) |
+
+## Appendix: opaque C pointer encoding
+
+A common practice in C library is to expose a set of functions, including a
+constructor, a destructor and several setters, to manipulate an opaque C
+pointer. Instead of directly provide Hack version of these functions, a `shape`
+or `vec` corresponds to the opaque C pointer should be expose, and the underlying C
+struct should be internally created on demand from the `shape` or `vec`, when
+another underlying C function requires the opaque C pointer.
```

Comment:
similarly, this section does not work generally for *opaque* pointers, despite that being the title. It assumes it is generally being used as a datastructure, rather than as a pointer to resource or something else.

Reply:
I used the opaque C pointer because this man page describe `posix_spawnattr_t` as an opaque data type: https://linux.die.net/man/3/posix_spawnattr_init

Anyway I removed the term `opaque`, because other opaque C pointers might not be short-lived, e.g. `FILE *`.